[vincent-maillou/qttools#161] [vincent-maillou/quatrex#27] Add wave function solver (QTBM)

almaeder opened the pull request:
**Issue by [vetschn](https://github.com/vetschn)**
_Originally opened as [https://github.com/vincent-maillou/qttools/pull/161](https://github.com/vincent-maillou/qttools/pull/161)_

This PR was originally in qttools repository (branch: add_cuDSS).

Original PR description:

----

This add direct sparse solvers for QTBM.

----

**Issue by [maurodossena](https://github.com/maurodossena)**
_Originally opened as [https://github.com/vincent-maillou/quatrex/pull/27](https://github.com/vincent-maillou/quatrex/pull/27)_

This PR was originally in quatrex repository (branch: Maurodossena/test qtbm).

Original PR description:

----

Added QTBM and some examples

---

## Tasks
- [x] Move away from run scripts; properly include QTBM in the CLI
- [x] k-point grid definition in the config instead of in the QTBM instantiation
- [x] Add an example for the carbon nanotube
- [ ] ~~Output options for orbital/atom resolved~~
- [x] Slight clean up of print statements and commented out code

maurodossena committed:
Added QTBM and atomic chain example

maurodossena committed:
Added contact class and fixed parallelization

maurodossena committed:
Implemented computation of multiple contacts transmissions, added effective mass 4 terminal device example

maurodossena committed:
Implemented transmission and DOS per slab

maurodossena committed:
Fixed slab indeces, added comparison with OMEN and QTsolver

maurodossena committed:
Added spill over corrections for DOS with S matrix and fixed bug affecting sparsity pattern

maurodossena committed:
Added comments

maurodossena committed:
Fixed obc call

maurodossena committed:
Added contact construction with permutations to avoid flipping

maurodossena committed:
Fixed spill over correction and cleaned up the code

maurodossena committed:
Fixed bug in case of block_section = 1

maurodossena committed:
Fixed small bug happening with CPU (index is not a scalar)

maurodossena committed:
Changed spsolve to splu

maurodossena committed:
Added nvidia cuDSS python binding

maurodossena committed:
Added cuDSS binding and added contact sorting method

maurodossena committed:
Normalize injection phase in Spectral class

maurodossena committed:
Integrate MUMPS solver as an alternative to CUDSS in QTBM class

maurodossena committed:
Fixed MUMPS calls

alexnick83 committed:
Added commented-out code for writing directly to the system matrix's buffer.

maurodossena committed:
Removed conflict, still to test sparse matrix fix

maurodossena committed:
Impented OBC batches, resolved conflicts, still to test sparse matrices fix from Alex

maurodossena committed:
On site update for sparse matrices, and new spill over corrections

maurodossena committed:
Rewrote cuDSS bindings to support sym fact reuse, and added recursion relation of Bloch matrix

maurodossena committed:
Some cleanup

maurodossena committed:
Some cleanup

maurodossena committed:
Cleanup

maurodossena committed:
Refactored QTBM to expose Contact and Device class. Implemented periodic self energies

vetschn committed:
Formatting and make the linter happy.

maurodossena committed:
Fixed all examples, changed hamiltonian label (3 index), fixed bugs with contact direction

maurodossena committed:
Commented a check that was causing issues

vetschn committed:
Initial refactoring of Device and QTBM classes

update method names, enhance initialization, and improve contact handling. More stuff to follow...

vetschn committed:
More refactoring of QTBM and Device classes.

No more contact/slab/orbital input files, instead define things in the config file. Simplified a few computations and changed some other stuff here and there.

Docstrings still need to be adjusted

vetschn committed:
Refactor slab index computations.

Use linspace and digitize to determine which atoms/orbitals are in which slab.

vetschn committed:
:robot: Have the AI write a bunch of docstrings

vetschn committed:
First draft of wave function solver interface, implementations (cuDSS, LU, MUMPS) and corresponding tests.

Everything now follows a predefined interface. MUMPS and cuDSS are able to reuse factorizations.

vetschn committed:
Refactor wave function solver: replace LU with SuperLU and update default parameters in cuDSS and MUMPS.

vetschn committed:
Refactor solver configuration and integrate solver selection in QTBM class.

vetschn committed:
Enhance cuDSS solver: add multithreading support and update operand reset options.

vetschn committed:
Add explicit reset method for right-hand side in cuDSS solver.

Also update tests to explicitely test for the case where the rhs changes number of columns.

vetschn committed:
Deprecate the original cuDSS binding in favor of the nvmath version.

vetschn committed:
Changes for GPU computations.

maurodossena committed:
Added return into _eig_cupy and added coloumns removal in full

maurodossena committed:
Fixed bug inside contact affecting the upscaling of SE

almaeder committed:
Refactored system reduction for full eigenvalue solving

almaeder commented:
is it possible to write a test for the injection vector? not sure what to test there

almaeder committed:
Unify energy grid creation between scba and qtbm

almaeder committed:
Added access to reduce sparsity in full solver

almaeder committed:
Refactoring OBC for qtbm

maurodossena committed:
Refectored QTBM (Added K-point grid, removed slabs, added current and electron charge computation)

maurodossena committed:
Fixed a bug when number of injected mode was equal to 1

maurodossena committed:
Fixed compatibility issue with .xyz files from ASE, and forced electron_charge_orb to be real

vetschn committed:
Fix a bug in the accessing of the nevp config settings.

vetschn committed:
Fix device class initialization without explicitely passing compute_config.

vetschn commented:
Some tasks i think should be done before merging:
- [ ] Move away from run scripts; properly include QTBM in the CLI
- [ ] k-point grid definition in the config instead of in the QTBM instantiation
- [ ] Add an example for the carbon nanotube
- [ ] Output options for orbital/atom resolved
- [ ] Slight clean up of print statements and commented out code

vetschn committed:
Memoizer 'enable' keyword has been replaced by a 'mode' literal.

No need to even check for this in qtbm.

vetschn committed:
Move the kpoint grid/shift definitions into the config.

Now these values are part of the config in the 'device' section. Also moved the qtbm batch size configuration into a separate config section and removed the unused code relating to the slabs.

maurodossena committed:
Added hole concentration and neutrality level

Kosheira1 committed:
Apply specific nvmath changes

almaeder committed:
Fine grained control and fixed tests

almaeder committed:
Non existing parameter

maurodossena committed:
Fixed bug inside nvmath eig binding (complex64 instead of complex128)

vetschn committed:
Add `formalism` parameter to QuatrexConfig and refactor run logic for QTBM and SCBA.

Now we dispatch QTBM/SCBA depending on this new required `formalism` parameter.

vetschn committed:
Update reference and CLI tests to reflect changes in the CLI.

vetschn committed:
Fixed inconsistent naming.

vetschn committed:
By default we should be targeting all apis, not just blas.

vetschn committed:
Suppress some superfluous/debug prints.

vetschn committed:
Carbon Nanotube QTBM example

maurodossena committed:
Used physical constants instead of a numeric value + restored all mode filtering

maurodossena committed:
Implemented extra condition for injected modes + computation of contact bands

maurodossena committed:
Add BandStructure calc, reduced memory consuption. Changed order of contact orbitals

maurodossena committed:
Added kernels to perform in-place operation to the system matrix (to avoid copies)

maurodossena committed:
Added back in the system size reduction in full

Due to rebasing the system size reduction was removed.
Also, due to the removal of the left keyword,
the system size reduction is now simplified.

maurodossena committed:
Added back a test for the system size reduction

maurodossena committed:
Fixed rebase conflicts for spectral

Additionally, added/removed the changes to the keywords
for the spectral solver in QTBM

almaeder commented:
Before merging this PR, we should do:

- [x] Setting up an integration test.
- [ ] Small refactors to get a similar structure to the current SCBA.

maurodossena committed:
Added QTBM atomic chain example and output writting.

This adds a new example for qtbm with the atomic chain.
Also the output writting was added, but needs to be cleaned up.
Not every output should be written out.

maurodossena committed:
Added reference test for the atomic chain and qtbm.

This added the reference test for transmission and DOS.
Also, the ouput files are renamed (remove ->).

maurodossena committed:
Changed to superlu for the example

maurodossena committed:
Added qtbm si_bulk test

almaeder committed:
Renaming variables in full nevp solver

This is done for clarity.

almaeder committed:
Removed format option from distributed_load and refactored _init_hamiltonian

If a different sparse format is needed, then transformation is done outside.
In _init_hamiltonian loading logic is not changed, but logic for orthogonal basis.

almaeder commented on tests/qttools/wave_function_solver/test_solve.py:
```diff
@@ -0,0 +1,147 @@
+    a = sparse.random(n, n, density=0.1, format=format, dtype=float)
+    a += 1j * sparse.random(n, n, density=0.1, format=format, dtype=float)
+    a += sparse.diags([2.0] * n, format=format, dtype=float)
+    b = xp.ones((n, m), dtype=xp.complex128, order=order)
```
Why is the test RHS only ones?

vetschn replied:
This should also work with random values i believe, so no real reason for this. Is one preferable over the other one?

almaeder commented on src/quatrex/core/quatrex_config.py:
```diff
@@ -362,13 +374,13 @@ class ElectronConfig(BaseModel):
         """Sets the left and right Fermi levels if not already set."""
         if (self.left_fermi_level is None) != (self.right_fermi_level is None):
-            raise ValueError(
+            warnings.warn(
```
I assume this was changed since QTBM uses the fermi levels from the contacts and doesnt set left/righ. Fermi levels should be refactored, but another quick fix would be to move this validator to QuatrexConfig and raise depending on the formalism. @vetschn what is your opinion?

vetschn replied:
>  Fermi levels should be refactored, but another quick fix would be to move this validator to QuatrexConfig and raise depending on the formalism.

I'm with you on this 👍 Fermi level should eventually be set per contact (so i guess they should end up in the `ContactConfig` for both SCBA and QTBM) and then validated depending on the transport formalism and the requested output (to compute Landauer-Büttiker current in QTBM you need the contacts' chemical potentials 💀 )

vetschn replied:
We'll deal with this when dealing with #187

almaeder commented on src/quatrex/core/subsystem.py:
```diff
@@ -92,7 +92,10 @@ def _configure_nevp(
+            return Full(
+                eig_compute_location=compute_config.nevp.eig_compute_location,
+                reduce=compute_config.nevp.reduce_sparsity,
+            )
```
Full in SCBA does currently not support the system size reduction. This will be added in a separate PR.

almaeder committed:
Small refactor fo the device class

Renaming overlap and hamiltonian matrices.
Enable both atom and orbital resolved potential by checking
the shape after loading the potential.
Removed the unused atoms_list and atom_potential attribut.

almaeder committed:
Moved monkhorst pack to a separate file

almaeder committed:
Initial contact refactor

Renaming, optimizing init functions

almaeder committed:
Refactor periodic transverse grid determination

almaeder committed:
Further device refactor

_get_atoms_transverse_sorted
_init_periodic_transverse_repetitions
_reorder_atoms

almaeder committed:
Start of the init hamiltonian matrix refactor

Separate function and using itertools

almaeder committed:
Refactored _init_hamiltonian_overlap_matrices

alexnick83 commented on examples/carbon-nanotube-dist/quatrex_config.toml:
```diff
@@ -1,18 +1,20 @@
 simulation_dir = "."
 
+formalism = "negf" # "negf" or "qtbm"
```
Isn't QTBM just NEGF without scattering? Shouldn't this be QTBM vs SCBA?

alexnick83 replied:
Do we actually need this variable? Couldn't we just check if there is an SCBA or QTBM entry in the configuration file?

almaeder replied:
Better naming would be "scba"/'qtbm" or "negf"/"wf".

vetschn replied:
I believe the naming logic was that in QTBM we never actually compute a Green's function, but in NE**GF** we do so, so the two are distinct from that point.

But I agree that checking whether there is a scba/qtbm entry in the config would likely be a better solution :+1:

almaeder replied:
I would say checking if entry exist for scba/qtbm is also not great since it doesnt allow to nicely switch between different type of simulations. For example, I had a setup which manipulates programmatically the default config and dumps the result to a new config file which results in both scba and qtbm entries existing.

alexnick83 replied:
However, the presence of both SCBA and QTBM is erroneous (unless we want to switch between those models using the same config file, which sounds error-prone). @almaeder maybe check #162 where I am writing with tomli-w to do the same, but I have the option to automatically skip defaults (and also decide exatly which keys should be written).

alexnick83 commented on examples/carbon-nanotube-qtbm/quatrex_config.toml:
```diff
@@ -0,0 +1,60 @@
+[device]
+transport_direction = "x"
+kpoint_grid = [1, 1, 1]
+num_orbitals_per_atom = { "X" = 1 }
```
Isn't this provided by the input data? Why do we need this?

almaeder replied:
The input data provides us with the hopping matrices (`hamiltonian_x_y_z.npz`),
but not for which points to solve.
Full hamiltonian in the system matrix would be ~`hamiltonian_0_0_0 + e^{iky}*hamiltonian_0_1_0 + e^{-iky}*hamiltonian_0_-1_0`. I believe k points in SCBA #205 are done in the same way.

We should also be able to directly input the full summed-up Hamiltonian, but this is for a subsequent PR.

alexnick83 replied:
I mean the `num_orbitals_per_atom`. To rephrase, why didn't we need this dictionary in SBCA?

almaeder replied:
Ah, it isn't currently needed in the SCBA since we have no notion of atoms.
The positions (`.xyz` file) and the potential are atome resolved qtbm, while the `grid.npy` is orbital resolved, I believe.
Instead, it should be possible to operate in a purely orbital picture, but some extra preprocessing would be needed, @maurodossena, right?

vetschn replied:
The QTBM inputs are now very different from the SCBA (they need to be coerced together asap). There is an `.xyz` file that contains the atom positions in the device. In CP2K (and other codes) different atoms can have different numbers of orbitals depending on the basis set (so maybe 13 orbitals per carbon atom but another number for molybdenum). The elements of the corresponding Hamiltonian are for each orbital.  

The main reason to keep track of `num_orbitals_per_atom` is to then sum together charges (that we compute per orbital) into charge per atom, which will be needed for Poisson.

We have not needed this in SCBA yet, because we always dealt with Wannier functions and not atom-centered basis functions. We should also have this for SCBA.

maurodossena replied:
> Ah, it isn't currently needed in the SCBA since we have no notion of atoms. The positions (`.xyz` file) and the potential are atome resolved qtbm, while the `grid.npy` is orbital resolved, I believe. Instead, it should be possible to operate in a purely orbital picture, but some extra preprocessing would be needed, @maurodossena, right?

Moving from an atoms picture to a pure orbital picture could be done with a very fast preprocessing step. I don't see any problem with that

vetschn replied:
Opened a new issue #214 where we will then work on this after merging the branch. Hope this is okay :)

alexnick83 commented on src/qttools/kernels/inplace/cupy/__init__.py:
```diff
@@ -0,0 +1,24 @@
+
+
+from qttools.kernels.inplace.cupy.inplace_add import (
+    add_kernel_comp,
```
Since all these kernels are operating in-place (I assume that is the reason for writing a custom implementation; cupy doesn't do in-place and uses extra memory), maybe call them `iadd` and `isub` to follow the same naming as the Python standard library?

alexnick83 replied:
@copilot can you identify those kernels and rename them appropriately? The renaming should be the following (but please validate):
- add_kernel_real/comp -> iadd_kernel_real/comp
- sub_kernel_real/comp -> isub_kernel_real/comp
- add_OBC_inplace -> iadd_OBC
- sub_OBC_inplace -> isub_OBC

alexnick83 commented on src/qttools/kernels/inplace/cupy/inplace_add.py:
```diff
@@ -0,0 +1,83 @@
+    "add_real_func",
+)
+
+add_OBC_inplace = cp.RawKernel(
```
Maybe interesting for @michael-wuethrich

alexnick83 commented on src/qttools/kernels/linalg/eig.py:
```diff
@@ -118,6 +134,144 @@ def _eig_cupy(
+    # TODO: this will not work for real A
+    # and the cupy binding should be followed
+
+    # TODO: not recommended to create/destroy handle every time
```
Maybe make the handle a singleton and initialize it when you import cusolverDn.

almaeder replied:
We could also get the handle from CuPy since it should be interoperable.

alexnick83 replied:
True. We need to discuss how much we want to use nvmath. We likely want to use cupy when functionality is equivalent because we can hope that eventually there will be parity with rocm/hip.

almaeder replied:
This was included by @Kosheira1 since the compilation from source to get eig from cupy was problematic.
With newer versions of cupy, our own nvmath eig should be removed.

almaeder replied:
Since the pre-release CuPy v14.0.0rc1 already contains EIG, I suggest removing the nvmath eig before merging. We shouldn't maintain our own binding.

vetschn replied:
I also prefer getting rid of our own custom wrapper for `cusolverDn.xgeev`. I don't expect that the one in the `cupy-14rc` has any meaningful difference in performance.

Since we check in `qtools.kernels.linalg.eig` whether `cupy` has `eig`, we already raise a `ValueError` in any case if cupy < v14, so no need to modify the `pyproject.toml`.

**Edit:**
Added a comment about this [here](https://github.com/quatrex/quatrex/issues/185#issuecomment-3748586395). We will keep it here until there is an official release of cupy 14.

alexnick83 commented on src/qttools/kernels/linalg/eig.py:
```diff
@@ -118,6 +134,144 @@ def _eig_cupy(
+
+    try:
+        # Query workspace size
+        lwork_device, lwork_host = cusolverDn.xgeev_buffer_size(
```
I am wondering whether the buffer-size calculation methods are slow enough that it would make sense to cache the result.

alexnick83 commented on src/qttools/nevp/full.py:
```diff
@@ -20,22 +22,95 @@ class Full(NEVP):
     use_pinned_memory : bool, optional
         Whether to use pinnend memory if cupy is used.
         Default is `True`.
+    a_sparsity : tuple[NDArray, ...] | None, optional
```
This seems quite expensive. Why not use CSC format, which allows you to find zero columns with `indptr[1:] - indptr[:-1]`?

vetschn replied:
See issue #215

alexnick83 commented on src/qttools/nevp/full.py:
```diff
@@ -73,12 +148,29 @@ def __call__(self, a_xx: tuple[NDArray, ...]) -> tuple[NDArray, NDArray]:
         A[:, : B.shape[0], : B.shape[1]] -= B
 
+        # Concatenate and delete
+        if self.reduce:
```
I think this needs some work. Is A sparse but with a different sparsity pattern than the NEVP coefficients?

almaeder replied:
It has a sparsity which can be determined from the coefficients:
<img width="864" height="514" alt="image" src="https://github.com/user-attachments/assets/7c19eac9-7a42-44c5-a1ff-bac70610023e" />
(Sascha Brück et. al)

vetschn replied:
Concluding: If A is sparse, then modes corresponding to zero columns are zero/infinite (which we do not care about). The sparsity is different than the coefficients, but can be computed from them. (@almaeder approves this message)

alexnick83 commented on src/qttools/obc/spectral.py:
```diff
@@ -489,22 +476,28 @@ def __call__(
+                )
+
+                # Compute surface phi
+                phi_surface.append(
```
Isn't the left-side matmul with `a_ji` (see line 504 on dev) missing here? Is that handled elsewhere now?

almaeder replied:
It was changed which quantities is  returned since qtbm needed both the injection vectors and the surface phi.
Now the injection vectors are outside compute instead of returning dependent quantities.

alexnick83 commented on src/qttools/wave_function_solver/__init__.py:
```diff
@@ -0,0 +1,8 @@
+# Copyright (c) 2025 ETH Zurich and the authors of the qttools package.
```
@vincent-maillou think if there is some intersection here with @ohcomely 's work. For example, if we identify three contacts as a significant bottleneck, it would be interesting to do a study of the performance of different solvers in the actual QTBM workflow.

alexnick83 commented on src/qttools/wave_function_solver/cudss.py:
```diff
@@ -0,0 +1,234 @@
+from qttools.profiling import Profiler
+from qttools.wave_function_solver.solver import WFSolver
+
+try:
```
nvmath is imported separately for eig and cuDSS. Consider adding a higher-level import similar to xp, since I expect we'll use it more and more.

vetschn replied:
Considered, but for when we use more nvmath since eig is just temporary, but cudss is forever.

alexnick83 commented on src/qttools/wave_function_solver/cudss.py:
```diff
@@ -0,0 +1,234 @@
+
+# Possible multithreading libraries for cuDSS in descending order of
+# preference.
+_mtlayer_libs = [
```
This looks like it will need a lot of maintenance. If nvmath is not handling it automatically, we should consider an abstraction for proper cuDSS installation (is it not part of the cudatoolkit?).

vetschn replied:
Unfortunately, at least when i was putting this together, it did not automatically detect an installed multithreading lib, and it ended up running single-threaded. The default seems to be to pass `multithreading_lib=None` into the `DirectSolver` instance (see docs [here](https://docs.nvidia.com/cuda/nvmath-python/0.5.0/host-apis/sparse/generated/nvmath.sparse.advanced.DirectSolverOptions.html#nvmath.sparse.advanced.DirectSolverOptions.multithreading_lib)).

cuDSS is not part of the cudatoolkit, it had to be installed seperately. I think depending on how `nvmath-python` is installed (conda/pip) it may have tried to also install it, but i would need to check again what the actual behavior is there

alexnick83 replied:
Then we update this as cuDSS evolves (I imagine that they will eventually provide a more robust solution).

alexnick83 commented on src/qttools/wave_function_solver/cudss.py:
```diff
@@ -0,0 +1,234 @@
+    def _explicitely_reset_b(self, b: NDArray) -> None:
+        """Resets the right-hand side in the direct solver.
+
+        This is a workaround for the current limitation in nvmath where
```
Do we (always) need this functionality? Is there logic surrounding this method to avoid calling it when not necessary?

vetschn replied:
in QTBM we solve `m @ x = b`, the symbolic factorization of `m` stays the same for all energies (so `direct_solver.plan()`, which is the slow part on CPU, should be called only once per run). The number of injection vectors (which sets the number of columns in `b`) is determined during the OBC calculation and can be different from one energy to the next. That's the reason for doing this "explicit reset", allocating new `b` and `x` with different shapes.

An option completely avoiding this would be to somehow find an upper bound for the number of injection vectors over all energies. There could potentially be a way to determine this from the contact band structure, but that would need some work and validation.

A compromise could be to check if the number of injection vectors stays the same from one energy to the next. If it does, explicitly reset, and implicitly reset if it doesn't. We could even sort the order of energies after the OBC computation and before the solve, to group the number of injection vectors and minimize the explicit resets

alexnick83 commented on src/qttools/wave_function_solver/cudss.py:
```diff
@@ -0,0 +1,234 @@
+        ds.rhs_strides = strides
+
+        ds.result_shape = shape
+        # NOTE: I do not know what these two lines do.
```
I read this as cuDSS supports a syntax for solving multiple linear systems with a single `A@x=b` solver statement, where A, x, and b are arrays of matrices/vectors. The rhs array may not necessarily have the same shape as the lhs; presumably, NumPy broadcasting rules apply.

vetschn replied:
This makes a lot of sense 👍 Thanks!

alexnick83 commented on src/qttools/wave_function_solver/mumps.py:
```diff
@@ -0,0 +1,90 @@
+valid_orderings = ["amd", "amf", "scotch", "pord", "metis", "qamd", "auto"]
+
+
+class MUMPS(WFSolver):
```
I would consider, in the future, discussing with @vincent-maillou and @ohcomely the potential integration with https://github.com/vincent-maillou/SolverBench (currently private), which will ultimately include installation scripts and bindings for solvers like MUMPS supporting Alps and some x86-based clusters. For now, just adding a few notes saying that this is experimental and please figure out yourselves how to install MUMPS (and METIS OR SCOTCH), and the python-mumps bindings suffice.

almaeder committed:
Refactored the coupling matrix creation

alexnick83 commented on src/qttools/wave_function_solver/solver.py:
```diff
@@ -0,0 +1,31 @@
+# Copyright (c) 2024 ETH Zurich and the authors of the qttools package.
```
2025, I guess. We should also start changing this to 2025-2026, but we can do a replacement and commit all files in another PR.

alexnick83 commented on src/quatrex/cli/main.py:
```diff
@@ -47,47 +47,63 @@ def version_callback(value: bool):
-        Path to the quatrex configuration file.
-    compute_config_path : Path | None, optional
-        Path to the compute configuration file, by default None.
+    quatrex_config : QuatrexConfig
```
I suggest you accept both a configuration object and a file (as previously). Or at the very least, let's sync on this because I also added something similar in the [better-batching PR](https://github.com/quatrex/quatrex/pull/172/changes/19f26dee511a7e7bedfd96ad6ac27382676d5948), needed for testing.

alexnick83 replied:
I see that you moved the file-handling to the CLI call, but not everything will go through there. Tests may (need to) use the `run` method directly. Of course, we could also have the tests handle the files or call some utility method; to discuss in more detail.

vetschn replied:
I suggest we keep this as is for the time being. In the integration tests we indeed go through the `run` function whenever configs should be parsed.

alexnick83 commented on src/quatrex/cli/main.py:
```diff
@@ -47,47 +47,63 @@ def version_callback(value: bool):
-        compute_config = ComputeConfig()
 
-    quatrex_config = parse_quatrex_config(quatrex_config_path)
+def _run_negf(quatrex_config, compute_config):
```
This is SCBA, right? Both are NEGF. Or let's clarify the naming.

alexnick83 commented on src/quatrex/device/contact.py:
```diff
@@ -0,0 +1,1199 @@
+from quatrex.core.quatrex_config import OBCConfig
+
+
+class Contact:
```
Is this generalizable to all NEFG models? Probably TODO for another PR.

almaeder replied:
Does currently not generalize due to the following:
- Simple contact description of left/right not possible or description through start/end index.
- Constructs OBC inputs without a possible having a self-energy.

vetschn replied:
See #187

alexnick83 commented on src/quatrex/device/contact.py:
```diff
@@ -0,0 +1,1199 @@
+
+        return coordinates
+
+    def _init_hamiltonian_overlap_matrices(self) -> int:
```
Why is this here? How does it differ from the existing utilities for creating Hamiltonians from primitive unit cells?

almaeder replied:
Difference is that the unit cells are "automatically"  discovered from the position and the hamiltonian elements.
There is no "unit_cells.npy" input which allows to have a periodic contact with an inhomogenous device,
but fully periodic device coming from "unit_cells.npy"  are  not  supported.

alexnick83 replied:
I see. If I understand correctly, this is necessary whenever you have a non-homogeneous device and need periodic contacts. I suppose this is independent of the model (i.e., we could run SCBA on such a device). In that case, I would try to move those to input utilities or whatever applies better.

almaeder replied:
Agree, but we should move the input utilities to another subfolder and separate different kinds.
Should we make the above move in this PR?
What I suggested should be better another one.

vetschn replied:
Maybe more cohesive to do this with #187

alexnick83 commented on src/quatrex/device/contact.py:
```diff
@@ -0,0 +1,1199 @@
+from quatrex.core.compute_config import NEVPConfig
+from quatrex.core.kpoints import monkhorst_pack
+from quatrex.core.quatrex_config import OBCConfig
+
```
I feel that this file has way too many things that go quite beyond the main theme (a Contact class). I would leave a big refactor (that also tries to unify the QTBM- and SCBA-related utilities) for another PR, but we should at least split the utilities.

almaeder replied:
Agree, the discovery of periodicity should be handled separately and
the contact should only know of the indices.
Then, the contact blocks are sliced during the `compute_boundary` call.

maurodossena replied:
I agree, the construction of contacts could be separated by the periodicity discovery

alexnick83 commented on src/quatrex/device/device.py:
```diff
@@ -0,0 +1,368 @@
+                flush=True,
+            )
+
+    def _init_hamiltonian(self) -> None:
```
This appears to load full Hamiltonians, and there is no option to construct them from primitive unit cells. However, primitive unit cells (UC) are mentioned in the contacts file, so I am bit confused on the input requirements of QTBM.

almaeder replied:
See https://github.com/quatrex/quatrex/pull/5#discussion_r2683981571.

almaeder committed:
Refactored _compute_band_structure

Use eigvalsh from qttools and the
compute config variables from the band edge config.

alexnick83 commented on src/quatrex/grid/kpoints.py:
```diff
@@ -0,0 +1,25 @@
+# Copyright (c) 2026 ETH Zurich and the authors of the quatrex package.
```
Is this synced with the SCBA k-points PR?

almaeder replied:
Kinda, both are using Monkhorst pack to generate the kpoint grid.

alexnick83 commented on src/quatrex/core/qtbm.py:
```diff
@@ -0,0 +1,1123 @@
+}
+
+
+def kr_mat_mul(m: NDArray, a: NDArray, vect: NDArray) -> NDArray:
```
This probably belongs in qttools.

alexnick83 commented on src/quatrex/core/qtbm.py:
```diff
@@ -0,0 +1,1123 @@
+
+    # 2. Apply 'a' to the first dimension (axis 0)
+    # tensordot(a, phi, axes=1) is like a @ phi along the first axis
+    temp = np.tensordot(a, vect_3d, axes=1)
```
Surely cupy has tensordot? Any reason this is happening on the CPU only?

almaeder replied:
This could be a bug since the inputs seem to be created with `xp`.
Changed it to `xp`. @maurodossena could you verify on the GPU?

maurodossena replied:
Good catch! I just copied this function from a prototype code and I forgot to convert `np` to `xp`. It should work also with cupy, I'll test that.

alexnick83 commented on src/quatrex/core/qtbm.py:
```diff
@@ -0,0 +1,1123 @@
+    return res_simple
+
+
+def allocate_sys_mat(
```
This needs a GPU implementation. Maybe we should go through the logic.

maurodossena replied:
I'll be happy to discuss a possible GPU implementation!

alexnick83 commented on src/quatrex/core/qtbm.py:
```diff
@@ -0,0 +1,1123 @@
+
+def allocate_sys_mat(
+    ham: dict, ovl: dict, boundary_SE_indexes: list[NDArray]
+) -> list[sparse.csr_matrix]:
```
Is the output a list? It doesn't seem so.

alexnick83 commented on src/quatrex/core/qtbm.py:
```diff
@@ -0,0 +1,1123 @@
+    return system_matrix
+
+
+def compute_update_indeces_sparse(
```
indices? (spelling)

alexnick83 commented on src/quatrex/core/qtbm.py:
```diff
@@ -0,0 +1,1123 @@
+    return xp.array(update_indices)
+
+
+def compute_update_indeces_dense(
```
indices? (spelling)

alexnick83 commented on src/quatrex/core/qtbm.py:
```diff
@@ -0,0 +1,1123 @@
+    res_simple = res_simple.transpose(0, 2, 1).reshape(-1, vect.shape[1], order="F")
+
+    return res_simple
+
```
This file seems to be full of stuff that needs a good sparse implementation, especially on the GPU.

alexnick83 commented on src/quatrex/core/qtbm.py:
```diff
@@ -0,0 +1,1123 @@
+        )
+
+        # Compute the current from all the k dependent transmissions
+        for n_t in range(self.num_transmissions):
```
Maybe this needs another compute-observables method?

alexnick83 reviewed (changes requested):
Overall, the code is well written. However, I see red flags in performance (to be addressed by working with students in subsequent PRs), and I am not convinced about the current hierarchy. At the very least, a lot of the methods should be moved to qttools as appropriate. We also need to discuss the overall workflow and how it aligns with SCBA (to be addressed in a subsequent PR). I would avoid any further development in this PR, aside from moving things around. It is already too big as-is.

almaeder committed:
Refactored contact matrix construction and injection mode upscaling in the Contact class

almaeder committed:
Addressing comments

Renaming formalism negf to scba, keeping scba naming.
Moving kron product to qttools linalg.
Moving preferred_matrix_type to wave_function_solver.

almaeder committed:
Kronecker product file

vetschn commented on src/qttools/utils/inplace_utils.py:
Some docstrings missing in this file (e.g. `sub_inplace_OBC`)

almaeder commented on src/quatrex/core/contact.py:
```diff
@@ -0,0 +1,1091 @@
+            device.quatrex_config.electron.obc, device.compute_config.nevp
+        )
+
+        self._compute_band_structure()
```
Now, every rank does the same contact band structure calculation.
The band structure is currently only used in `qtbm.py`, where it is saved to file.
I would suggest removing the contact band structure calculation for now and adding in a separate PR a "formalism" that only does this without having to do QTBM.

maurodossena replied:
Agree! I added that just as a test for a specific calculation I was doing

almaeder replied:
I removed it.

almaeder committed:
Refactor compute_observables of the QTBM class

Additionally, change variable naming and remove
the shape counting variable from compute_observables
since it can be infered from the injection vector output.

almaeder committed:
Refactor QTBM, moved current and charge computations to separate methods.

Renamed compute_observables to _compute_observables.

almaeder committed:
Moved add compute_update_indices_sparse and compute_update_indices_dense to inplace_utils.py

maurodossena commented on src/qttools/obc/spectral.py:
If I remember correctly, the function should support batchsize > 1. Maybe this ValueError comes from an older version of the function

almaeder replied:
Agree, this could be due to the rebasing we done.

vetschn replied:
We'll keep this restriction for now, as we have not tested the injection vector calculation with batchsize > 1 yet. (Adding a TODO in the code)

almaeder committed:
Remove band structure from QTBM

Copilot committed:
Rename inplace kernels to follow Python naming conventions (iadd/isub prefix)

Co-authored-by: alexnick83 <31545860+alexnick83@users.noreply.github.com>

alexnick83 commented:
@copilot this PR introduces several new files under `src/quatrex/core`. We want to organize them a bit better. We would like to group the files `contact.py` and `device.py` under a `device` submodule. We also want to group `energies.py` and `kpoints.py` under a `grid` submodule. Can you move the files and adapt the code as appropriate?

Copilot committed:
Reorganize core files into device and grid submodules

Co-authored-by: alexnick83 <31545860+alexnick83@users.noreply.github.com>

vetschn committed:
Renamed the formalism options.

The formalism keyword was not super descriptive. Now we use "wf" / "negf".

vetschn committed:
Refactored the "inplace" kernels.

Mainly shifted around kernels and wrapping functions and renamed things. Now numba and cupy backend have a common interface. Kernels get dispatched based on the present array module, like they should.

vetschn committed:
Moved the 'device' and 'grid' subpackages out of the 'core' subpackage to the quatrex top level.

Also adjusted imports accordingly.

Feels a bit more natural this way. There is an argument to be had about whether 'core' should even exist as it is not very descriptive :skull:

vetschn committed:
Renaming several attributes and variables.

We (@almaeder and I) renamed a bunch of variable to more accurately reflect how they are used in the code. Particularly we renamed what was previously `size` in the contact definition to `lattice_vectors`, which meant we had to change the example configs.

vetschn committed:
Formatting (reordered imports)

vetschn committed:
Add TODO comment regarding batchsize restriction for injection vector calculation.

This will have to be taken care of in another PR. Specifically, a test should cover this.

vetschn committed:
Change docs to reflect that QTBM does not track band edges.

This was previously used in some contact band structure computations. Not needed anymore.

vetschn committed:
Renamings and smallish refactorings.

We tried to come up with some better names for several attributes. While going through we also ended up changing how the variable previously known as `orbital_indices_per_repetition` is storing the `unit_cell_orbital_indices`. Added/modified some docstrings as well.

alexnick83 reviewed (approved):
It looks good to me now, based on what we discussed as targets for completing the PR.

vetschn committed:
Refactor the system matrix allocation in QTBM.

Some simplification, we no longer construct lists and then pass those lists around but rather we operate directly on the `device.contacts` list.

vetschn committed:
Refactored how transmissions and contact currents are stored.

Before they were stored across multiple separate values, now they are dicts, where the keys are (`contact_in`, `contact_out`). The label is constructed just for the outputs.

vetschn committed:
Removed electron/hole charge observables.

These are the excess charges integrated from the neutrality level up. There is only one charge neutrality level across the device. These are not really meaningful observables and were unused anyway.

vetschn committed:
Removed unused attributes, and further refactorings of the observable calculations.

Now the observable calculation is split up into transmission/LDOS. The LDOS is now also a dictionary with contacts as keys, rather than an array with a contact index.

vetschn committed:
Renamed the inplace kernel parameters.

Now unpack the tuple in the kernels directly.
